### ISSUE TITLE
CLI Observations and Usage Recommendations.

### ISSUE THREAD

**Issue by bradleykdowning** [ibm-mas/installer#200]
Through repeated use of the CLI tool, I have come to the belief that this is a nice tool.  It would be better if:

1. I had a list of the tool contents and possible paths before I had to discover and make every mistake possible to know HOW to use the tool.
2. There is no instruction anywhere on the possible routes through the CLI.
3. The fact that I cannot install predict WITHOUT doing Manage is silly. In other words YES it must be installed FIRST but if I have already done so please provide me the ability to create a new pipeline run without forcing me back through the Manage install pipeline run.... ugh...
4. I really should install CPD on a separate track PRIOR to installing IOT or Manage.  Without CPD the DBs end up on islands and I have to do a lot of work to gain visibility within CPD UI in my cloud account.  In earlier version it was part and parcel of the UI and I had visibility as part of install.  Please bring this back. 
5. Is this team aware of the work begin done over here:   [https://github.com/cloud-native-toolkit/operator-masauto](https://github.com/cloud-native-toolkit/operator-masauto).  It occurs to me that the work is do a degree complimentary and overlapping.   The idea of an operator on the front end of the gun is very attractive for those who prefer to have a point-and-click experience.  Further the choices and options in the CLI would be nice to see in the front end of the Operator during the set up of the operator.  Dunno if that can be done.
7. Finally, If the CLI had the ability to incorporate choices into the execution of the pipeline runs (and moreover a recommendation of the order of the pipeline runs) then I could choose to install through either the CLI or the pipeline runs in the OCP UI, OR the Automation Operator the various applications in a manner of my choosing all with the same user experience vis-a-vis the order and choices.  

In other words when I install Maximo application suite I really should be presented with the same set of options in the same order whether it is through a CLI tool or an operator tool. 

IBM has enough knowledge on HOW the install should take place... don't make the user guess.

**Comment by durera** [comment]
> I had a list of the tool contents and possible paths before I had to discover and make every mistake possible to know HOW to use the tool.

I might be misinterpreting the question, but is this what you're looking for: e.g. https://ibm-mas.github.io/cli/commands/mirror-images/ & https://ibm-mas.github.io/cli/commands/install/

> There is no instruction anywhere on the possible routes through the CLI.

Can you elaborate, in terms of the interactive install the goal was that the messages presented convey enough information to guide the user through the prompts to not require a seperate document.  Were there specific parts where it wasn't clear what input was required?

> I really should install CPD on a separate track PRIOR to installing IOT or Manage. Without CPD the DBs end up on islands

We found the db2u standalone operator to have a smaller footprint, be easier to manage updates and upgrades, and provide an all round more reliable experience which is why we have focused on it rather than Db2 in CPD as the "automation provided" option.  This allows us to only bring CPD into the picture when a select few applications are installed.  The CLI supports "bring you own database" now, which would allow someone who wants to use Db2 inside CPD to do so.  

BYO options for all other dependencies are planned as well.

> Is this team aware of the work begin done over here: https://github.com/cloud-native-toolkit/operator-masauto.  It occurs to me that the work is do a degree complimentary and overlapping. The idea of an operator on the front end of the gun is very attractive for those who prefer to have a point-and-click experience.

Yup, at some point there will be a GUI fronting the CLI, the basic idea is a small web application inside the CLI container that can be optionally started to provide a point and click interface to the same capability provided on the command line currently.

There are no plans to integrate CLI with the operator, our focus for alternative install/management of MAS is on native support for a GitOps approach to managing MAS on a cluster: https://github.com/ibm-mas/gitops/tree/poc

> The fact that I cannot install predict WITHOUT doing Manage is silly. In other words YES it must be installed FIRST but if I have already done so please provide me the ability to create a new pipeline run without forcing me back through the Manage install pipeline run.... ugh...

Once manage is installed, the second pipeline run will be a no-op and should complete almost instantly.  Similar to the rest of the pipeline stages that have already been completed.  Each step is designed to be idempotent, so if the desired state is already met then nothing changes and the step should complete after making no changes; it's basically just verifying that what you wanted is already there.

The same applies to all repeated install commands, each time to you are telling the pipeline "put my cluster into this final state" ... if you add a new app on a subsequent run you still need to be providing the same inputs for existing applications.  "Install over the top" is supported by the pipeline, but it's not a "modify existing install" action.  We could make the install aware of the starting state of the MAS instance to support it being a general purpose "modify", but that goes beyond the original intent/goals of the installer and would create an uncontainable amount of testing for each release as we would need to ensure that the install can cope with the myriad of possible starting states.

> Finally, If the CLI had the ability to incorporate choices into the execution of the pipeline runs (and moreover a recommendation of the order of the pipeline runs)

I don't follow sorry, the CLI provides choices for all inputs to the pipeline, and the stages of the pipeline are designed to run in a recommended order/pattern already.

**Comment by durera** [comment]
Closing due to inactivity, not sure if the response earlier in the year addressed everything.  If not feel free to reply further in this or a new issue.

**Comment by bradleykdowning** [comment]
Ha!  I never saw the comments back David.  NO wonder I never added stuff back.  Hahahaha..... oh geez.

BTW I have seen improvements in the CLI.  I have not actually touched it since July.  Recently (last week 18 October ) I saw a product demo with the CLI and It actually incorporates some answers to my concerns above.  So BRAVO great work!!

Insofar as the point on "no instructions" that still stands.  In my view it would be nice to have a document that previews the CLI.  With the preview, I am prepared for each of the questions in the interactive mode.  Without knowing what questions are coming up, I MUST pause and go and get the answer if I do not have all possible information for which I know not the question.  Since it has been a while, I am lacking on specifics (I know, I know... I should have the specific example.  I will get it to you.)

Insofar as the last comment.... I reread that and it like like I never finished the thought. Ignore.  

Overall Thanks for the repossess (albeit FOUR month later.... damn..... Hahahahahaha....) Great work David and Team!!
